core/genesis.go

```diff
@@ -245,30 +234,35 @@ func SetupGenesisBlock(
 	// when we start syncing from scratch, the last accepted block
 	// will be genesis block
 	if lastBlock == nil {
-		return newcfg, fmt.Errorf("missing last accepted block")
+		return newcfg, common.Hash{}, fmt.Errorf("missing last accepted block")
 	}
-
 	height := lastBlock.NumberU64()
 	timestamp := lastBlock.Time()
 	if skipChainConfigCheckCompatible {
 		log.Info("skipping verifying activated network upgrades on chain config")
 	} else {
 		compatErr := storedcfg.CheckCompatible(newcfg, height, timestamp)
-		if compatErr != nil && height != 0 && compatErr.RewindTo != 0 {
-			return newcfg, compatErr
+		if compatErr != nil && ((height != 0 && compatErr.RewindToBlock != 0) || (timestamp != 0 && compatErr.RewindToTime != 0)) {
+			return newcfg, stored, compatErr
 		}
 	}
-	rawdb.WriteChainConfig(db, stored, newcfg)
-	return newcfg, nil
+	// Don't overwrite if the old is identical to the new
+	if newData, _ := json.Marshal(newcfg); !bytes.Equal(storedData, newData) {
```

Comment:
We need to remove this check so that it writes the config to disk (including the upgrade config).